### PR TITLE
Streams

### PR DESCRIPTION
**Pull request by 1BAH**

### CODE REVIEW

File: src/main/java/com/github/fppt/jedismock/exception/WrongStreamKeyException.java [hunk 1]
```diff
@@ -0,0 +1,7 @@
+package com.github.fppt.jedismock.exception;
```
Comment (inponomarev) [inline src/main/java/com/github/fppt/jedismock/exception/WrongStreamKeyException.java:1]: all the other exceptions in this package extend `RuntimeException`. Any purpose of this being the checked exception?

File: pom.xml [hunk 2]
```diff
@@ -106,7 +107,7 @@
             <groupId>org.slf4j</groupId>
             <artifactId>slf4j-simple</artifactId>
             <version>2.0.9</version>
-            <scope>test</scope>
+<!--            <scope>test</scope>-->
```
Comment (inponomarev) [inline pom.xml:110]: why is this commented out? slf4j-simple is not intended to be a transitive dependency

Comment (1BAH) [reply pom.xml:110]: Hmm my fault - do not remember exactly the purpose of doing it..

File: src/main/java/com/github/fppt/jedismock/exception/WrongStreamKeyException.java (hunk 1 above)
Comment (1BAH) [reply src/main/java/com/github/fppt/jedismock/exception/WrongStreamKeyException.java:1]: I've made it a checked exception to ensure that when keys are being constructed from slices wrong data cases are checked. Otherwise it is a high probability of being missed.

File: src/main/java/com/github/fppt/jedismock/datastructures/streams/StreamId.java [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+        this(0, 0);
+    }
+
+    /* ID from long numbers for PRIVATE API usage */
+    public StreamId(long firstPart, long secondPart) {
+        this.firstPart = firstPart;
+        this.secondPart = secondPart;
+    }
+
+    public StreamId(String key) throws WrongStreamKeyException {
```
Comment (inponomarev) [inline src/main/java/com/github/fppt/jedismock/datastructures/streams/StreamId.java:37]: This is a much shorter version that does the same:

```java
        Matcher matcher = Pattern.compile("(\\d+)(?:-(\\d+))?").matcher(key);
        if (matcher.matches()) {
            try {
                firstPart = parseUnsignedLong(matcher.group(1));
                secondPart = matcher.group(2) == null ? 0 : parseUnsignedLong(matcher.group(2));
            } catch (NumberFormatException e) {
                throw new WrongStreamKeyException(INVALID_ID_ERROR);
            }
        } else {
            throw new WrongStreamKeyException(INVALID_ID_ERROR);
        }
```

File: src/main/java/com/github/fppt/jedismock/datastructures/streams/StreamId.java [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+                throw new WrongStreamKeyException(INVALID_ID_ERROR);
+        }
+    }
+
+    public StreamId(Slice slice) throws WrongStreamKeyException {
+        this(slice.toString());
+    }
+
+    public StreamId compareToZero() throws WrongStreamKeyException {
+        if (secondPart == 0 && firstPart == 0) {
```
Comment (inponomarev) [inline src/main/java/com/github/fppt/jedismock/datastructures/streams/StreamId.java:75]: AFAICS the only use of this in source code is catching and rethrowing the result

Please rename it and rewrite it to `boolean isZero()` so that exception is not used to control flow

File: src/main/java/com/github/fppt/jedismock/datastructures/streams/StreamId.java [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+            }
+
+            --first;
+        }
+
+        return new StreamId(first, second);
+    }
+
+    @Override
+    public int compareTo(StreamId other) {
```
Comment (inponomarev) [inline src/main/java/com/github/fppt/jedismock/datastructures/streams/StreamId.java:113]: So if `compareUnsigned != 0` you are comparing it twice. Why not cache the result of the first comparison in a variable?

Why we are bothering with `unsigned` throughout this class at all? Can we consider this values as signed so that we utilize only 1/2 of `long` range? I believe we discussed this earlier on one of our calls.


In particular, this could be rewritten as 

```java
return Comparator.comparing(StreamId::getFirstPart)
                .thenComparing(StreamId::getSecondPart).compare(this, other);
```

File: src/main/java/com/github/fppt/jedismock/datastructures/streams/StreamId.java [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+        return Slice.create(toString());
+    }
+
+    @Override
+    public String toString() {
+        return toUnsignedString(firstPart) + "-" + toUnsignedString(secondPart);
+    }
+
+    @Override
+    public boolean equals(Object o) {
```
Comment (inponomarev) [inline src/main/java/com/github/fppt/jedismock/datastructures/streams/StreamId.java:129]: This requires an `EqualsVerifier` test

File: src/test/java/com/github/fppt/jedismock/datastructures/streams/StreamIdTests.java [hunk 7]
```diff
@@ -0,0 +1,259 @@
...
+                WrongStreamKeyException.class,
+                zero::compareToZero,
+                "ERR The ID specified in XADD must be greater than 0-0"
+        );
+    }
+
+    @Test
+    void zeroComparisonWithPositiveKeysTest() {
+        StreamId other = new StreamId(0, 1);
+        assertDoesNotThrow(() -> assertSame(other, other.compareToZero()));
```
Comment (inponomarev) [inline src/test/java/com/github/fppt/jedismock/datastructures/streams/StreamIdTests.java:24]: no need in `assertDoesNotThrow`. If you're worrying about unchecked exception, just declare it in a test method, and test will fail if exception is thrown. The use case for `assertDoesNotThrow` is for usage in  [soft asserts](https://eliasnogueira.com/soft-asserts-why-should-you-use-them-for-unit-and-integration-tests/), and here we don't use soft asserts.

Anyway, this class should be rewritten to use AssertJ

File: src/test/java/com/github/fppt/jedismock/datastructures/streams/StreamIdTests.java [hunk 8]
```diff
@@ -0,0 +1,259 @@
...
+
+        assertDoesNotThrow(() -> assertEquals(
+                new StreamId(-1234567891, -1234567891),
+                new StreamId(-1234567891, -1234567890).decrement()
+        ));
+    }
+
+    @Test
+    void incrementDecrementFuzzingTest() {
+        for (int i = 0; i < 1000; ++i) {
```
Comment (inponomarev) [inline src/test/java/com/github/fppt/jedismock/datastructures/streams/StreamIdTests.java:103]: Please do not write such tests. They are not real fuzzing tests, and you should not copy the approaches from TCL native tests in Redis (they are restricted there by the available tooling and are quite foolish "poor man's fuzzing").

Real fuzzing is more complex than this, it's a combination of well-thought selection of "bad", "corner-case" examples with tooling that makes any failure reproducible (if something fails, we must be able to know which example failed).

Fuzzing for Java is provided by [jqwik](https://jqwik.net/), but I'm in doubt whether we should pull  `jqwik` into this project. Just a parameterized test with a couple of examples will cover everything and will not bring any additional value.

Please, do not write unit tests like this, use parameterized tests instead for 2-3 examples including corner cases

**Review by inponomarev — CHANGES_REQUESTED** [review]

File: src/main/java/com/github/fppt/jedismock/datastructures/streams/StreamId.java (hunk 4 above)
Comment (1BAH) [reply src/main/java/com/github/fppt/jedismock/datastructures/streams/StreamId.java:75]: The idea of such implementation is that when in XADD we are trying to create StreamID:
```java
try {
     nodeId = compareWithTopKey(new StreamId(stream.replaceAsterisk(id)).compareToZero());
} catch (WrongStreamKeyException e) {
     return Response.error(e.getMessage());
}
```

constructor might throw an exception as well as `compareWithTopKey` method. So I've writen this way to simplify code and follow the same style

File: src/test/java/com/github/fppt/jedismock/datastructures/streams/StreamIdTests.java (hunk 7 above)
Comment (1BAH) [reply src/test/java/com/github/fppt/jedismock/datastructures/streams/StreamIdTests.java:24]: WrongStreamKeyException is a checked exception

Comment (inponomarev) [reply src/test/java/com/github/fppt/jedismock/datastructures/streams/StreamIdTests.java:24]: Yep it was my typo, I meant checked exception. Just make `zeroComparisonWithPositiveKeysTest() throws WrongStreamKeyException`. Again,  `assertDoesNotThrow` makes sense in the context of soft asserts only.

File: src/main/java/com/github/fppt/jedismock/datastructures/streams/StreamId.java (hunk 4 above)
Comment (inponomarev) [reply src/main/java/com/github/fppt/jedismock/datastructures/streams/StreamId.java:75]: Your `compareWithTopKey` and `compareToZero` methods (note the arbitrary usage of 'with' and 'to') actually return their argument or throw an exception. So, they are not transforming their argument but just validating it. For this scenario, in Java practice, methods in the form of `void validateFoo(...) throws Bar` are used. This code can be rewritten in a clearer way without shifting the responsibility of validation to the `StreamId` class (see code below). And even this variant is far from perfect. If we can throw fewer exceptions in Java, it's better to do so. Each exception carries a call stack (leading to heap & garbage collector overload and CPU time to collect the call stack). No compiler optimizations are made for exceptional branches of code -- intentionally.

```java
//package-private, not public, for testing purposes
public void validate(StreamId key) throws WrongStreamKeyException {
        if (key.isZero()) { //no RMStream is needed at this point, so let's check it first
            throw new WrongStreamKeyException(ZERO_ERROR);
        }
        StreamId lastId = getStreamFromBaseOrCreateEmpty(params().get(0)).getLastId();
        if (key.compareTo(lastId) <= 0) {
            throw new WrongStreamKeyException(TOP_ERROR);
        }
    }


protected Slice response() {
   //...blah- blah...
        StreamId nodeId;
        try {
            nodeId = new StreamId(stream.replaceAsterisk(id));
            validate(nodeId);
        } catch (WrongStreamKeyException e) {
            return Response.error(e.getMessage());
        }
  //...
}
```

**Review by inponomarev — APPROVED** [review]
